lib/rucio/daemons/storage/consistency/actions.py

```diff
@@ -0,0 +1,805 @@
+                new_bad_replica = models.BadReplicas(scope=scope, name=name, rse_id=rse_id,
+                  reason=reason, state=status, account=issuer, bytes=size)
+                new_bad_replica.save(session=session, flush=False)
+                session.query(models.Source).filter_by(scope=scope, name=name,
+                  rse_id=rse_id).delete(synchronize_session=False)
+            else:
+                if already_declared:
+                    unknown_replicas.append('%s:%s %s' % (did['scope'], did['name'],
+                     'Already declared'))
+                else:
+                    unknown_replicas.append('%s:%s %s' % (did['scope'], did['name'],
+                     'Unknown replica'))
+        if str(status) == str(BadFilesStatus.BAD):
+            # For BAD file, we modify the replica state, not for suspicious
+            try:
+                # there shouldn't be any exceptions since all replicas exist
+                update_replicas_states(replicas, session=session)
+            except exception.UnsupportedOperation:
+                raise exception.ReplicaNotFound("One or several replicas don't exist.")
+    try:
+        session.flush()
+    except IntegrityError as error:
+        raise exception.RucioException(error.args)
+    except DatabaseError as error:
+        raise exception.RucioException(error.args)
+    except FlushError as error:
+        raise exception.RucioException(error.args)
+
+    return unknown_replicas
+### NOTE: declare_bad_file_replicas will be used directly from core/replica.py
+### when handling of DID is added there
+##########################################################################
+
+
+##############################
+### This is Igor's Stats class
+class Stats(object):
+
+    def __init__(self, path):
+        self.Path = path
+        self.Data = {}
+
+    def __getitem__(self, name):
+        return self.Data[name]
+
+    def __setitem__(self, name, value):
+        self.Data[name] = value
+        self.save()
+
+    def get(self, name, default=None):
+        return self.Data.get(name, default)
+
+    def update(self, data):
+        self.Data.update(data)
+        self.save()
+
+    def save(self):
+        try:
+            with open(self.Path, "r") as f:
+                data = f.read()
+        except:
+            #traceback.print_exc()
+            data = ""
+        data = json.loads(data or "{}")
+        data.update(self.Data)
+        open(self.Path, "w").write(json.dumps(data, indent=4))
+
+
+def write_stats(my_stats, stats_file, stats_key=None):
+    if stats_file:
+        stats = {}
+        if os.path.isfile(stats_file):
+            with open(stats_file, "r") as f:
+                stats = json.loads(f.read())
+        if stats_key:
+            stats[stats_key] = my_stats
+        else:
+            stats.update(my_stats)
+        open(stats_file, "w").write(json.dumps(stats))
+
+
+
+def cmp2dark(new_list="T2_US_Purdue_2021_06_18_02_28_D.list",
```

Comment:
these default parameters look like they come from a unittest?

Reply:
You are right - those defaults are not needed in the production version.
Removed.